File: app/Shared/ChainTopology.swift
import Foundation

enum ChainTopology {
    static func chainOctet(chainID: UUID) -> Int {
        let hex = chainID.uuidString.replacingOccurrences(of: "-", with: "")
        var hash: UInt32 = 2_166_136_261
        for byte in hex.utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 16_777_619
        }
        return 1 + Int(hash % 254)
    }

    static func overlayCIDR(chainID: UUID) -> String {
        "10.64.\(chainOctet(chainID: chainID)).0/24"
    }

    static func overlaySubnet(chainID: UUID) -> String {
        "10.64.\(chainOctet(chainID: chainID)).0"
    }

    static func overlayMask() -> String {
        "255.255.255.0"
    }

    static func listenPort(chainID: UUID) -> Int {
        7400 + chainOctet(chainID: chainID)
    }

    static func overlayAddr(chainID: UUID, index: Int) -> String {
        "10.64.\(chainOctet(chainID: chainID)).\(HopConstants.overlayNodeOctet + index)"
    }
}

struct TunnelConnectContext: Codable, Equatable {
    let chainID: UUID
    let hopperPort: Int
    let overlayCIDR: String
    let deviceID: UUID
}

File: app-android/app/src/main/java/com/aengix/hopper/model/ChainTopology.kt
package com.aengix.hopper.model

object ChainTopology {
    fun chainOctet(chainId: String): Int {
        val hex = chainId.replace("-", "")
        var hash = 2_166_136_261L
        for (ch in hex.toByteArray(Charsets.UTF_8)) {
            hash = hash xor ch.toLong()
            hash = (hash * 16_777_619L) and 0xFFFF_FFFFL
        }
        return (1 + (hash % 254)).toInt()
    }

    fun overlayCIDR(chainId: String): String = "10.64.${chainOctet(chainId)}.0/24"

    fun overlaySubnet(chainId: String): String = "10.64.${chainOctet(chainId)}.0"

    fun listenPort(chainId: String): Int = 7400 + chainOctet(chainId)

    fun overlayAddr(chainId: String, index: Int): String =
        "10.64.${chainOctet(chainId)}.${HopConstants.OVERLAY_NODE_OCTET + index}"
}
